File: Network.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import AS
import networkx as nx
import pylab

class Network:
  def __init__(self):
    self.as_list = {}
    self.addr_to_as = {}
    self.top_as_numbers = []
    self.graph = nx.DiGraph()

  def loadRelationFile(self, filename):
    for line in open(filename, 'r'):
      if line.startswith('#'):
        continue  # コメント行
      para = line.split('|')
      if len(para) != 3:
        continue  # 改行 or その他の関係の無い行

      node1 = int(para[0])
      node2 = int(para[1])
      connection_type = int(para[2])

      if node1 not in self.as_list:
        self.as_list[node1] = AS.AS(node1)
        self.graph.add_node(node1)
      if node2 not in self.as_list:
        self.as_list[node2] = AS.AS(node2)
        self.graph.add_node(node2)

      if (connection_type == -1):
        # transit connection, node1 provide node2
        provider = self.as_list[node1]
        customer = self.as_list[node2]
        provider.customerNodes[customer.as_number] = customer
        customer.providerNodes[provider.as_number] = provider
        # customer -> provider
        self.graph.add_edge(node2, node1, weight=1, label='transit')
      elif (connection_type == 0):
        # peer connection
        peer1 = self.as_list[node1]
        peer2 = self.as_list[node2]
        peer1.peerNodes[peer2.as_number] = peer2
        peer2.peerNodes[peer1.as_number] = peer1
        self.graph.add_edge(node1, node2, weight=0, label='peer')

    return len(self.as_list)

  def searchTopAS(self):
    for node in self.as_list.values():
      if len(node.providerNodes) == 0:
        self.top_as_numbers.append(node.as_number)
    
    return len(self.top_as_numbers)

  def exchangeRoutingTable(self):
    # トップレベルASから再帰的に下位ASのcustomerTableを統合
    for as_number in self.top_as_numbers:
      top_as = self.as_list[as_number]
      top_as.mergeCustomerTable()

    # 全ASでピアにcustomerTableの内容を通知
    for node in self.as_list.values():
      node.mergePeerTable()

    # トップレベルASから再帰的に下位ASにcustomerTable, peerTable,
    # providerTableの内容を通知
    for as_number in self.top_as_numbers:
      top_as = self.as_list[as_number]
      top_as.transit()

  def distributeNetAddr(self):
    seq = 1
    for node in self.as_list.values():
      networkAddr = seq << 16   # 上位16ビットがネットワーク部、下位16ビットがホスト部
      node.networkAddress = networkAddr
      node.netmask = 0xFFff0000
      seq += 1
      self.addr_to_as[networkAddr] = node.as_number

  def traceroute(self, src, dst, startTime=0, packets=3):
    delay = 13   # delay per router
    elapsed = 0 # ms
    src = AS.AS.strAddrToInt(src)
    dst = AS.AS.strAddrToInt(dst)
    mask = 0xFFff0000
    srcNet = src & mask
    dstNet = dst & mask

    src_as_number = self.addr_to_as[srcNet]
    dst_as_number = self.addr_to_as[dstNet]

    current = self.as_list[src_as_number]
    while True:
      # ホップ毎に指定されたパケット数を送信する
      for i in range(packets):
        elapsed += delay
        current.addLog(src, dst, startTime + elapsed)
      route = None
      if dst_as_number in current.customerTable:
        route = current.customerTable[dst_as_number]
      elif dst_as_number in current.peerTable:
        route = current.peerTable[dst_as_number]
      elif dst_as_number in current.providerTable:
        route = current.providerTable[dst_as_number]
      else:
        return -1
      
      nexthop = route["Nexthop"]
      if nexthop == -1:
        break
      else:
        current = self.as_list[nexthop]

    return elapsed

  def drawGraph(self):
    transit = [(u,v) for (u,v,d) in self.graph.edges(data=True) 
        if d['weight'] > 0.5]
    peer = [(u,v) for (u,v,d) in self.graph.edges(data=True) 
        if d['weight'] <= 0.5]
    pylab.figure(figsize=(4, 4))
    # pos = nx.spring_layout(self.graph)
    pos = nx.circular_layout(self.graph)
    # pos = nx.random_layout(self.graph)
    nx.draw_networkx_nodes(self.graph, pos, node_size=30, node_color='w')
    nx.draw_networkx_edges(self.graph, pos, edgelist=transit, width=1, edge_color='b')
    nx.draw_networkx_edges(self.graph, pos, edgelist=peer, arrows=False, edge_color='g', style='dashed')
    nx.draw_networkx_labels(self.graph, pos, font_size=20, font_color='r')
    pylab.xticks([])
    pylab.yticks([])
    pylab.savefig("network.png")
    pylab.show()

  def clearASLog(self):
    for node in self.as_list.values():
      node.clearLog()

File: Network.cpp
/*
 * Network.cpp
 * Copyright (C) 2015 takayuki <takayuki@Takayukis-MacBook-Pro.local>
 *
 * Distributed under terms of the MIT license.
 */

#include "AS.cpp"
#include <fstream>
#include <algorithm>
using namespace std;

class Network {
public:
  map<unsigned int, int> addrToAsNumber;
  map<int, AS *> asDict;
  vector<AS *> asList;
  vector<int> topAsNumbers;
  
  ~Network() {
    this->addrToAsNumber.clear();
    this->asDict.clear();
    this->topAsNumbers.clear();
    for (int i=0; i<asList.size(); i++) {
      delete asList[i];
      asList[i] = NULL;
    }
    this->asList.clear();
  }

  AS *createAS() {
    AS *a = new AS();
    asList.push_back(a);
    return a;
  }

  int loadFile(string &filename) {
    map<int, AS *>::iterator itr;
    string str;
    int i, node1, node2, connectionType;
    AS *provider, *customer, *peer1, *peer2;
    AS *node;

    ifstream fin(filename);
    if (!fin) {
      cout << "error : open input file" << endl;
      return -1;
    }

    fin.unsetf(ios::skipws);
    while (getline(fin, str)) {
      vector<string> para = AS::split(str, "|");
      if (para.size() != 3) {
        continue;
      }

      node1 = stoi(para[0]);
      node2 = stoi(para[1]);
      connectionType = stoi(para[2]);

      itr = asDict.find(node1);
      if (itr == asDict.end()) {
        node = createAS();
        *node = AS::AS(node1);
        asDict.insert(pair<int, AS *>(node1, node));
      }
      itr = asDict.find(node2);
      if (itr == asDict.end()) {
        node = createAS();
        *node = AS::AS(node2);
        asDict.insert(pair<int, AS *>(node2, node));
      }

      if (connectionType == -1) {
        itr = asDict.find(node1);
        provider = itr->second;
        itr = asDict.find(node2);
        customer = itr->second;
        provider->addCustomerAS(*customer);
        customer->addProviderAS(*provider);
      } else if (connectionType == 0) {
        itr = asDict.find(node1);
        peer1 = itr->second;
        itr = asDict.find(node2);
        peer2 = itr->second;
        peer1->addPeerAS(*peer2);
        peer2->addPeerAS(*peer1);
      }
      sort(asList.begin(), asList.end());
    }

    for (int i=0; i<asList.size(); i++) {
      asList[i]->init();
    }
    return 0;
  }

  vector<int> &searchTopAs() {
    map<int, AS *>::iterator itr = this->asDict.begin();
    while (itr != this->asDict.end()) {
      if (itr->second->providerNodes.size() == 0) {
        this->topAsNumbers.push_back(itr->first);
      }
      itr++;
    }
    sort(this->topAsNumbers.begin(), this->topAsNumbers.end());
    return this->topAsNumbers;
  }

  void exchangeRouteingTable() {
    map<int, AS *>::iterator itrDict;
    AS *node;
    int i;
    
    for (i=0; i<topAsNumbers.size(); i++) {
      cout << "TopAsNumbers:" << topAsNumbers[i] << endl;
      itrDict = this->asDict.find(topAsNumbers[i]);
      node = itrDict->second;
      node->mergeCustomerTable();
    }

    for (i=0; i<asList.size(); i++) {
      cout << "AllAs:" << asList[i]->asNumber << endl; 
      node = asList[i];
      node->mergePeerTable();
    }

    for (i=0; i<topAsNumbers.size(); i++) {
      cout << "TopAsNumbers:" << topAsNumbers[i] << endl;
      itrDict = this->asDict.find(topAsNumbers[i]);
      node = itrDict->second;
      node->transit();
    }
  }

  void distributeNetAddr() {
    map<int, AS *>::iterator itr = this->asDict.begin();
    AS *node;
    while (itr != this->asDict.end()) {
      node = itr->second;
      node->networkAddress = (unsigned int)(node->asNumber<<16);
      node->netmask = 0xFFff0000;
      this->addrToAsNumber.insert(
          pair<unsigned int, int>(node->networkAddress, node->asNumber));
      itr++;
    }
  }

  int traceroute(unsigned int src, unsigned int dst,
      unsigned int startTime=0, int packets=3) {
    unsigned int delay = 13;
    unsigned int elapsed = 0;
    unsigned int mask = 0xFFff0000;
    unsigned int srcNet, dstNet;
    int srcAsNumber, dstAsNumber, i, nexthop;
    AS *current;
    Route *r;
    map<unsigned int, int>::iterator itrAddr;
    map<int, AS *>::iterator itrDict;
    map<int, Route *>::iterator itrTable;
    
    srcNet = src & mask;
    dstNet = dst & mask;
    itrAddr = this->addrToAsNumber.find(srcNet);
    srcAsNumber = itrAddr->second;
    itrAddr = this->addrToAsNumber.find(dstNet);
    dstAsNumber = itrAddr->second;

    itrDict = this->asDict.find(srcAsNumber);
    current = itrDict->second;
    while (1) {
      for (i=0; i<packets; i++) {
        elapsed += delay;
        current->addLog(src, dst, startTime + elapsed);
      }
      if ((itrTable = current->customerTable.find(dstAsNumber))
            != current->customerTable.end()) {
        r = itrTable->second;
      } else if ((itrTable = current->peerTable.find(dstAsNumber))
          != current->peerTable.end()) {
        r = itrTable->second;
      } else if ((itrTable = current->providerTable.find(dstAsNumber))
          != current->providerTable.end()) {
        r = itrTable->second;
      } else {
        return -1;
      }
      nexthop = r->nexthop;
      if (nexthop == -1) {
        break;
      } else {
        itrDict = this->asDict.find(nexthop);
        current = itrDict->second;
      }
    }
    return elapsed;
  }

  void crearAsLog() {
    map<int, AS *>::iterator itr = this->asDict.begin();
    while (itr != this->asDict.end()) {
      itr->second->clearLog();
      itr++;
    }
  }

  void toString() {
    map<int, AS *>::iterator itr = asDict.begin();
    while (itr != asDict.end()) {
      itr->second->toString();
      itr->second->printLog();
      itr++;
    }
  }
};

int main() {
  string filename = "./topology/test.txt";
  Network myNet;
  myNet.loadFile(filename);
  myNet.searchTopAs();
  for (int i=0; i<myNet.topAsNumbers.size(); i++) {
    cout << "TOP AS : " << myNet.topAsNumbers[i] << endl;
  }
  myNet.exchangeRouteingTable();
  myNet.distributeNetAddr();
  myNet.toString();

  myNet.traceroute((7<<16 | 1), (14<<16 | 1), 0, 3);

  return 0;
}
